File: platform/processing-core/app/schemas/cases.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.cases import CaseCommentType, CaseKind, CasePriority, CaseStatus


class CaseCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: CaseKind
    entity_id: str | None = None
    kpi_key: str | None = None
    window_days: int | None = Field(default=None, ge=1)
    title: str | None = Field(default=None, max_length=160)
    priority: CasePriority = CasePriority.MEDIUM
    note: str | None = None
    explain: dict[str, Any] | None = None
    diff: dict[str, Any] | None = None
    selected_actions: list[dict[str, Any]] | None = None


class CaseUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: CaseStatus | None = None
    assigned_to: str | None = None
    priority: CasePriority | None = None


class CaseCommentCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    body: str = Field(min_length=1)


class CaseSnapshotOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    explain_snapshot: dict[str, Any]
    diff_snapshot: dict[str, Any] | None
    selected_actions: list[dict[str, Any]] | None
    note: str | None
    created_at: datetime


class CaseCommentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    author: str | None
    type: CaseCommentType
    body: str
    created_at: datetime


class CaseResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: int
    kind: CaseKind
    entity_id: str | None
    kpi_key: str | None
    window_days: int | None
    title: str
    status: CaseStatus
    priority: CasePriority
    created_by: str | None
    assigned_to: str | None
    created_at: datetime
    updated_at: datetime
    last_activity_at: datetime


class CaseListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[CaseResponse]
    total: int
    limit: int
    next_cursor: str | None = None


class CaseDetailsResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    case: CaseResponse
    latest_snapshot: CaseSnapshotOut | None
    comments: list[CaseCommentOut]
    snapshots: list[CaseSnapshotOut] | None = None


__all__ = [
    "CaseCommentCreateRequest",
    "CaseCommentOut",
    "CaseCreateRequest",
    "CaseDetailsResponse",
    "CaseListResponse",
    "CaseResponse",
    "CaseSnapshotOut",
    "CaseUpdateRequest",
]

File: frontends/client-portal/src/types/cases.ts
export type CaseKind = "operation" | "invoice" | "order" | "support" | "dispute" | "incident" | "kpi" | "fleet" | "booking";
export type CaseStatus = "TRIAGE" | "IN_PROGRESS" | "WAITING" | "RESOLVED" | "CLOSED";
export type CasePriority = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";
export type CaseQueue = "FRAUD_OPS" | "FINANCE_OPS" | "SUPPORT" | "GENERAL";
export type CaseSlaState = "ON_TRACK" | "WARNING" | "BREACHED";

export interface CaseSnapshot {
  id: string;
  explain_snapshot: Record<string, unknown>;
  diff_snapshot?: Record<string, unknown> | null;
  selected_actions?: { code: string; what_if?: Record<string, unknown> | null }[] | null;
  note?: string | null;
  created_at: string;
}

export interface CaseComment {
  id: string;
  author?: string | null;
  type: "user" | "system";
  body: string;
  created_at: string;
}

export interface CaseItem {
  id: string;
  tenant_id: number;
  kind: CaseKind;
  entity_type?: string | null;
  entity_id?: string | null;
  kpi_key?: string | null;
  window_days?: number | null;
  title: string;
  description?: string | null;
  status: CaseStatus;
  queue: CaseQueue;
  priority: CasePriority;
  escalation_level: number;
  first_response_due_at?: string | null;
  resolve_due_at?: string | null;
  sla_state?: CaseSlaState | null;
  client_id?: string | null;
  partner_id?: string | null;
  created_by?: string | null;
  assigned_to?: string | null;
  case_source_ref_type?: string | null;
  case_source_ref_id?: string | null;
  created_at: string;
  updated_at: string;
  last_activity_at: string;
}

export interface CaseListResponse {
  items: CaseItem[];
  total: number;
  limit: number;
  next_cursor?: string | null;
}

export interface CaseDetailsResponse {
  case: CaseItem;
  latest_snapshot?: CaseSnapshot | null;
  comments: CaseComment[];
  timeline: Array<{ status: CaseStatus; occurred_at: string }>;
  snapshots?: CaseSnapshot[] | null;
}
